daemon/tcp_listener.c

```diff
@@ -172,20 +172,25 @@ static void streambuf_listener_newconn(struct obj *p, socket_t *newsock, char *a
 	i.timer = streambuf_stream_timer;
 	i.obj = &s->obj;
 
-	if (poller_add_item(listener->poller, &i))
-		goto fail;
-
 	if (cb->newconn_func)
 		cb->newconn_func(s);
 
+	obj_hold(s);
+
 	mutex_lock(&listener->lock);
 	g_hash_table_insert(listener->streams, s, s); // hand over ref
 	mutex_unlock(&listener->lock);
 
+	if (poller_add_item(listener->poller, &i))
+		goto fail;
+
+	obj_put(s);
+
 	return;
 
 fail:
 	obj_put(s);
+	obj_put(s);
```

Comment:
This one should not be necessary. The `fail` label is only reached after `s` has been added to `streams` which hands over one reference. So you're left with only one reference to release.

These changes should be squashed into a single commit. I can do that and submit it if you prefer.